lib/grandpa/message_handler.go

```diff
@@ -91,6 +98,57 @@ func (h *MessageHandler) handleFinalizationMessage(msg *FinalizationMessage) (*C
 	return nil, nil
 }
 
+func (h *MessageHandler) verifyJustification(fm *FinalizationMessage) error {
+	// validate justification
+	justList := fm.Justification
+	sigCount := 0
+	for _, j := range justList {
+		// verify signature
+		msg, err := scale.Encode(&FullVote{
+			Stage: precommit,
+			Vote:  NewVote(j.Vote.hash, j.Vote.number),
+			Round: fm.Round,
+			SetID: h.grandpa.state.setID,
+		})
+		if err != nil {
+			return err
+		}
+
+		pk, err := ed25519.NewPublicKey(j.AuthorityID[:])
+		if err != nil {
+			return err
+		}
+
+		ok, err := pk.Verify(msg, j.Signature[:])
+		if err != nil {
+			return err
+		}
+
+		if !ok {
+			return ErrInvalidSignature
+		}
+
+		// verify authority in justification set
+		authFound := false
+		for _, auth := range h.grandpa.Authorities() {
+			if reflect.DeepEqual(auth.Key.AsBytes(), j.AuthorityID) {
+				authFound = true
+				break
+			}
+		}
+		if !authFound {
+			return ErrVoterNotFound
+		}
+		sigCount++
+	}
+
+	// confirm total # signatures >= 2/3 of number of voters
+	if !(sigCount >= (2/3)*len(h.grandpa.state.voters)) {
```

Comment:
you could use `h.grandpa.state.threshold()` instead of `(2/3)*len(h.grandpa.state.voters)`

Reply:
good idea, this also fixes the issue you pointed out with the test.